Build issues using GitLab CI

- Description of the bug
while building the vaadin flow from the GitLab CI using the production mode there are issues in connecting the yarnpkg.com which is perfectly pingable.

The --network-concurrency 1 should fix this issue, but there is no way to pass this parameter to the yarn. 

`
[INFO] Installing Yarn version v1.6.0
[INFO] Downloading https://github.com/yarnpkg/yarn/releases/download/v1.6.0/yarn-v1.6.0.tar.gz to /builds/login-ops/hlp-resource-manager/target/cache/yarn-v1.6.0.tar.gz
[INFO] No proxies configured
[INFO] No proxy was configured, downloading directly
May 15, 2018 12:54:50 PM org.apache.http.client.protocol.ResponseProcessCookies processCookies
WARNING: Invalid cookie header: "Set-Cookie: logged_in=no; domain=.github.com; path=/; expires=Sat, 15 May 2038 12:54:50 -0000; secure; HttpOnly". Invalid 'expires' attribute: Sat, 15 May 2038 12:54:50 -0000
[INFO] Unpacking /builds/login-ops/hlp-resource-manager/target/cache/yarn-v1.6.0.tar.gz into /builds/login-ops/hlp-resource-manager/target/node/yarn
[INFO] Installed Yarn locally.
[INFO] Running 'yarn install' in /builds/login-ops/hlp-resource-manager/target
[INFO] yarn install v1.6.0
[INFO] [1/4] Resolving packages...
[INFO] [2/4] Fetching packages...
[ERROR] error An unexpected error occurred: "https://registry.yarnpkg.com/babel-helpers/-/babel-helpers-6.24.1.tgz: connect EHOSTUNREACH 104.16.59.173:443".
`
- Minimal reproducible example

Using the standard showcase project to reproduce. Tested with flow-maven-plugin / 1.0.0.beta7 on vaadin flow 10.0.0.beta10. Setting the yarn version to the latest v1.6.0 does not help.

- Expected behavior
The build should work (an option to pass yarn parameters might help to mitigate this sort of issues)
- Actual behavior
The build cannot be performed.

--- a/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/common/FrontendToolsManager.java
+++ b/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/common/FrontendToolsManager.java
@@ -28,6 +28,11 @@ import java.util.Set;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import com.github.eirslett.maven.plugins.frontend.lib.FrontendPluginFactory;
 import com.github.eirslett.maven.plugins.frontend.lib.InstallationException;
 import com.github.eirslett.maven.plugins.frontend.lib.NodeInstaller;
@@ -35,10 +40,6 @@ import com.github.eirslett.maven.plugins.frontend.lib.ProxyConfig;
 import com.github.eirslett.maven.plugins.frontend.lib.TaskRunnerException;
 import com.github.eirslett.maven.plugins.frontend.lib.YarnInstaller;
 import com.google.common.collect.ImmutableMap;
-import org.apache.commons.io.FileUtils;
-import org.apache.commons.io.IOUtils;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 /**
  * Entity to operate frontend tools to transpile files.
@@ -46,15 +47,20 @@ import org.slf4j.LoggerFactory;
  * Transpilation is based on <a href="https://github.com/gulpjs/gulp">gulp</a>
  * and <a href="https://github.com/Polymer/polymer-build">polymer-build</a>.
  * <p>
- * Polymer tools are used since they are the only ones capable of bundling html files along with css and js ones and providing helper method for processing web components.
+ * Polymer tools are used since they are the only ones capable of bundling html
+ * files along with css and js ones and providing helper method for processing
+ * web components.
  * <p>
- * Apart from gulp and polymer-build, <a href="https://github.com/yarnpkg/yarn">yarn</a> and <a href="https://github.com/nodejs/node">node</a>
- * are used to get and launch the tools required.
+ * Apart from gulp and polymer-build,
+ * <a href="https://github.com/yarnpkg/yarn">yarn</a> and
+ * <a href="https://github.com/nodejs/node">node</a> are used to get and launch
+ * the tools required.
  *
  * @author Vaadin Ltd.
  */
 public class FrontendToolsManager {
-    private static final Logger LOGGER = LoggerFactory.getLogger(FrontendToolsManager.class);
+    private static final Logger LOGGER = LoggerFactory
+            .getLogger(FrontendToolsManager.class);
 
     private final FrontendPluginFactory factory;
     private final File workingDirectory;
@@ -64,39 +70,65 @@ public class FrontendToolsManager {
 
     /**
      * Prepares the manager.
-     * @param workingDirectory          the directory to install and process files in, not {@code null}
-     * @param es5OutputDirectoryName    the name for the directory to put transpiled ES5 files into, not {@code null}
-     * @param es6OutputDirectoryName    the name for the directory to put minified ES6 files into, not {@code null}
-     * @param frontendDataProvider      the source of the files required by the {@link FrontendToolsManager}, not {@code null}
+     *
+     * @param workingDirectory
+     *            the directory to install and process files in, not
+     *            {@code null}
+     * @param es5OutputDirectoryName
+     *            the name for the directory to put transpiled ES5 files into,
+     *            not {@code null}
+     * @param es6OutputDirectoryName
+     *            the name for the directory to put minified ES6 files into, not
+     *            {@code null}
+     * @param frontendDataProvider
+     *            the source of the files required by the
+     *            {@link FrontendToolsManager}, not {@code null}
      */
     public FrontendToolsManager(File workingDirectory,
-                                String es5OutputDirectoryName, String es6OutputDirectoryName, FrontendDataProvider frontendDataProvider) {
-        FlowPluginFileUtils.forceMkdir(Objects.requireNonNull(workingDirectory));
-        this.factory = new FrontendPluginFactory(workingDirectory, workingDirectory);
+            String es5OutputDirectoryName, String es6OutputDirectoryName,
+            FrontendDataProvider frontendDataProvider) {
+        FlowPluginFileUtils
+                .forceMkdir(Objects.requireNonNull(workingDirectory));
+        this.factory = new FrontendPluginFactory(workingDirectory,
+                workingDirectory);
         this.workingDirectory = workingDirectory;
-        this.es5OutputDirectoryName = Objects.requireNonNull(es5OutputDirectoryName);
-        this.es6OutputDirectoryName = Objects.requireNonNull(es6OutputDirectoryName);
-        this.frontendDataProvider = Objects.requireNonNull(frontendDataProvider);
+        this.es5OutputDirectoryName = Objects
+                .requireNonNull(es5OutputDirectoryName);
+        this.es6OutputDirectoryName = Objects
+                .requireNonNull(es6OutputDirectoryName);
+        this.frontendDataProvider = Objects
+                .requireNonNull(frontendDataProvider);
     }
 
     /**
-     * Installs tools required for transpilation.
-     * Tools installed are:
+     * Installs tools required for transpilation. Tools installed are:
      * <ul>
      * <li><a href="https://github.com/nodejs/node">node</a></li>
      * <li><a href="https://github.com/yarnpkg/yarn">yarn</a></li>
      * <li><a href="https://github.com/gulpjs/gulp">gulp</a></li>
      * </ul>
-     * Additionally, {@literal gulpfile.js} file that is used for installation and transpilation via {@literal gulp}.
+     * Additionally, {@literal gulpfile.js} file that is used for installation
+     * and transpilation via {@literal gulp}.
      *
-     * @param proxyConfig proxy config to use when downloading frontend tools, not {@code null}
-     * @param nodeVersion node version, not {@code null}
-     * @param yarnVersion yarn version, not {@code null}
-     * @throws IllegalStateException if dependency installation fails
-     * @throws UncheckedIOException  if supplementary file creation fails
+     * @param proxyConfig
+     *            proxy config to use when downloading frontend tools, not
+     *            {@code null}
+     * @param nodeVersion
+     *            node version, not {@code null}
+     * @param yarnVersion
+     *            yarn version, not {@code null}
+     * @param networkConcurrency
+     *            maximum number of concurrent network requests
+     *
+     * @throws IllegalStateException
+     *             if dependency installation fails
+     * @throws UncheckedIOException
+     *             if supplementary file creation fails
      */
-    public void installFrontendTools(ProxyConfig proxyConfig, String nodeVersion, String yarnVersion) {
-        LOGGER.info("Installing required frontend tools to '{}'", workingDirectory);
+    public void installFrontendTools(ProxyConfig proxyConfig,
+            String nodeVersion, String yarnVersion, int networkConcurrency) {
+        LOGGER.info("Installing required frontend tools to '{}'",
+                workingDirectory);
         Objects.requireNonNull(proxyConfig);
         Objects.requireNonNull(nodeVersion);
         Objects.requireNonNull(yarnVersion);
@@ -104,99 +136,147 @@ public class FrontendToolsManager {
         createFileFromTemplateResource("package.json", Collections.emptyMap());
         createFileFromTemplateResource("yarn.lock", Collections.emptyMap());
         try {
-            factory.getNodeInstaller(proxyConfig)
-                    .setNodeVersion(nodeVersion)
-                    .setNodeDownloadRoot(NodeInstaller.DEFAULT_NODEJS_DOWNLOAD_ROOT)
+            factory.getNodeInstaller(proxyConfig).setNodeVersion(nodeVersion)
+                    .setNodeDownloadRoot(
+                            NodeInstaller.DEFAULT_NODEJS_DOWNLOAD_ROOT)
                     .install();
-            factory.getYarnInstaller(proxyConfig)
-                    .setYarnVersion(yarnVersion)
-                    .setYarnDownloadRoot(YarnInstaller.DEFAULT_YARN_DOWNLOAD_ROOT)
+            factory.getYarnInstaller(proxyConfig).setYarnVersion(yarnVersion)
+                    .setYarnDownloadRoot(
+                            YarnInstaller.DEFAULT_YARN_DOWNLOAD_ROOT)
                     .install();
-            factory.getYarnRunner(proxyConfig, null)
-                    .execute("install", Collections.emptyMap());
+            StringBuilder args = new StringBuilder("install");
+            if (networkConcurrency >= 0) {
+                args.append(" --network-concurrency ");
+                args.append(networkConcurrency);
+            }
+            factory.getYarnRunner(proxyConfig, null).execute(args.toString(),
+                    Collections.emptyMap());
         } catch (InstallationException | TaskRunnerException e) {
-            throw new IllegalStateException("Failed to install required frontend dependencies", e);
+            throw new IllegalStateException(
+                    "Failed to install required frontend dependencies", e);
         }
     }
 
-    private void createFileFromTemplateResource(String templateResourceName, Map<String, String> replacements) {
-        try (InputStream resource = getClass().getClassLoader().getResourceAsStream(templateResourceName)) {
+    private void createFileFromTemplateResource(String templateResourceName,
+            Map<String, String> replacements) {
+        try (InputStream resource = getClass().getClassLoader()
+                .getResourceAsStream(templateResourceName)) {
             if (resource == null) {
-                throw new IllegalStateException(String.format("File '%s' is not found in application resources", templateResourceName));
+                throw new IllegalStateException(String.format(
+                        "File '%s' is not found in application resources",
+                        templateResourceName));
             }
 
-            String resourceContents = IOUtils.toString(resource, StandardCharsets.UTF_8);
-            for (Map.Entry<String, String> replacement : replacements.entrySet()) {
-                resourceContents = resourceContents.replaceFirst(Pattern.quote(replacement.getKey()), replacement.getValue().replace("\\", "/"));
+            String resourceContents = IOUtils.toString(resource,
+                    StandardCharsets.UTF_8);
+            for (Map.Entry<String, String> replacement : replacements
+                    .entrySet()) {
+                resourceContents = resourceContents.replaceFirst(
+                        Pattern.quote(replacement.getKey()),
+                        replacement.getValue().replace("\\", "/"));
             }
 
-            FileUtils.write(new File(workingDirectory, templateResourceName), resourceContents, StandardCharsets.UTF_8);
+            FileUtils.write(new File(workingDirectory, templateResourceName),
+                    resourceContents, StandardCharsets.UTF_8);
         } catch (IOException e) {
-            throw new UncheckedIOException(String.format("Failed to create a file in '%s'", workingDirectory), e);
+            throw new UncheckedIOException(String.format(
+                    "Failed to create a file in '%s'", workingDirectory), e);
         }
     }
 
     /**
-     * Transpiles files in the directory provided, copying them to the working directory.
-     * Creates a new output directory, in working directory, with two separate directories for ES5 (transpiled) and ES6 (optimized) files.
+     * Transpiles files in the directory provided, copying them to the working
+     * directory. Creates a new output directory, in working directory, with two
+     * separate directories for ES5 (transpiled) and ES6 (optimized) files.
      * Additionally, creates supplementary files, used by transpilation process.
      *
-     * @param es6SourceDirectory   the directory to get ES6 files from, not {@code null}
-     * @param outputDirectory      the directory to put transpiled files into, created if absent
-     * @param skipEs5              whether to skip the transpilation step or not
+     * @param es6SourceDirectory
+     *            the directory to get ES6 files from, not {@code null}
+     * @param outputDirectory
+     *            the directory to put transpiled files into, created if absent
+     * @param skipEs5
+     *            whether to skip the transpilation step or not
      * @return generated output directory path
-     * @throws IllegalStateException    if transpilation fails
-     * @throws IllegalArgumentException if es6SourceDirectory is not a directory or does not exist
-     * @throws UncheckedIOException     if output directory creation fails or other {@link IOException} occurs
+     * @throws IllegalStateException
+     *             if transpilation fails
+     * @throws IllegalArgumentException
+     *             if es6SourceDirectory is not a directory or does not exist
+     * @throws UncheckedIOException
+     *             if output directory creation fails or other
+     *             {@link IOException} occurs
      */
-    public Map<String, File> transpileFiles(File es6SourceDirectory, File outputDirectory, boolean skipEs5) {
+    public Map<String, File> transpileFiles(File es6SourceDirectory,
+            File outputDirectory, boolean skipEs5) {
         LOGGER.info("Processing frontend files from '{}'", es6SourceDirectory);
         try {
             FileUtils.forceMkdir(Objects.requireNonNull(outputDirectory));
         } catch (IOException e) {
-            throw new UncheckedIOException(String.format("Failed to create output directory '%s'", outputDirectory), e);
+            throw new UncheckedIOException(
+                    String.format("Failed to create output directory '%s'",
+                            outputDirectory),
+                    e);
         }
 
         if (!Objects.requireNonNull(es6SourceDirectory).isDirectory()) {
-            throw new IllegalArgumentException(String.format("es6SourceDirectory '%s' is not a directory or does not exist", es6SourceDirectory));
+            throw new IllegalArgumentException(String.format(
+                    "es6SourceDirectory '%s' is not a directory or does not exist",
+                    es6SourceDirectory));
         }
 
         ImmutableMap.Builder<String, String> gulpFileParameters = new ImmutableMap.Builder<String, String>()
-                .put("{es6_source_directory}", es6SourceDirectory.getAbsolutePath())
+                .put("{es6_source_directory}",
+                        es6SourceDirectory.getAbsolutePath())
                 .put("{target_directory}", outputDirectory.getAbsolutePath())
                 .put("{es5_configuration_name}", es5OutputDirectoryName)
                 .put("{es6_configuration_name}", es6OutputDirectoryName)
-                .put("{bundle}", Boolean.toString(frontendDataProvider.shouldBundle()))
-                .put("{minify}", Boolean.toString(frontendDataProvider.shouldMinify()))
-                .put("{hash}", Boolean.toString(frontendDataProvider.shouldHash()))
-                .put("{shell_file}", frontendDataProvider.createShellFile(workingDirectory))
-                .put("{fragment_files}", combineFilePathsIntoString(frontendDataProvider.createFragmentFiles(workingDirectory)));
-        createFileFromTemplateResource("gulpfile.js", gulpFileParameters.build());
+                .put("{bundle}",
+                        Boolean.toString(frontendDataProvider.shouldBundle()))
+                .put("{minify}",
+                        Boolean.toString(frontendDataProvider.shouldMinify()))
+                .put("{hash}",
+                        Boolean.toString(frontendDataProvider.shouldHash()))
+                .put("{shell_file}",
+                        frontendDataProvider.createShellFile(workingDirectory))
+                .put("{fragment_files}",
+                        combineFilePathsIntoString(frontendDataProvider
+                                .createFragmentFiles(workingDirectory)));
+        createFileFromTemplateResource("gulpfile.js",
+                gulpFileParameters.build());
 
         Map<String, File> transpilationResults = new HashMap<>();
         try {
-            factory.getGulpRunner().execute("build_es6", Collections.emptyMap());
-            addTranspilationResult(transpilationResults, outputDirectory, es6OutputDirectoryName);
+            factory.getGulpRunner().execute("build_es6",
+                    Collections.emptyMap());
+            addTranspilationResult(transpilationResults, outputDirectory,
+                    es6OutputDirectoryName);
 
             if (!skipEs5) {
-                factory.getGulpRunner().execute("build_es5", Collections.emptyMap());
-                addTranspilationResult(transpilationResults, outputDirectory, es5OutputDirectoryName);
+                factory.getGulpRunner().execute("build_es5",
+                        Collections.emptyMap());
+                addTranspilationResult(transpilationResults, outputDirectory,
+                        es5OutputDirectoryName);
             }
         } catch (TaskRunnerException e) {
-            throw new IllegalStateException("Transpilation with gulp has failed", e);
+            throw new IllegalStateException(
+                    "Transpilation with gulp has failed", e);
         }
 
         return transpilationResults;
     }
 
     private String combineFilePathsIntoString(Set<String> fragmentFiles) {
-        return fragmentFiles.stream().map(fileName -> "'" + fileName + "'").collect(Collectors.joining(", "));
+        return fragmentFiles.stream().map(fileName -> "'" + fileName + "'")
+                .collect(Collectors.joining(", "));
     }
 
-    private void addTranspilationResult(Map<String, File> transpilationResults, File outputDirectory, String configurationName) {
-        File configurationOutput = FileUtils.getFile(outputDirectory, configurationName);
+    private void addTranspilationResult(Map<String, File> transpilationResults,
+            File outputDirectory, String configurationName) {
+        File configurationOutput = FileUtils.getFile(outputDirectory,
+                configurationName);
         if (!configurationOutput.isDirectory()) {
-            throw new IllegalStateException(String.format("Unable to find transpilation result directory at '%s'", configurationOutput));
+            throw new IllegalStateException(String.format(
+                    "Unable to find transpilation result directory at '%s'",
+                    configurationOutput));
         }
         transpilationResults.put(configurationName, configurationOutput);
     }

--- a/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/maven/PackageForProductionMojo.java
+++ b/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/maven/PackageForProductionMojo.java
@@ -24,7 +24,6 @@ import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
 
-import com.github.eirslett.maven.plugins.frontend.lib.ProxyConfig;
 import org.apache.maven.artifact.DependencyResolutionRequiredException;
 import org.apache.maven.execution.MavenSession;
 import org.apache.maven.plugin.AbstractMojo;
@@ -39,6 +38,7 @@ import org.apache.maven.settings.crypto.DefaultSettingsDecryptionRequest;
 import org.apache.maven.settings.crypto.SettingsDecrypter;
 import org.apache.maven.settings.crypto.SettingsDecryptionResult;
 
+import com.github.eirslett.maven.plugins.frontend.lib.ProxyConfig;
 import com.vaadin.flow.plugin.common.AnnotationValuesExtractor;
 import com.vaadin.flow.plugin.common.FlowPluginFileUtils;
 import com.vaadin.flow.plugin.common.FrontendDataProvider;
@@ -91,6 +91,9 @@ public class PackageForProductionMojo extends AbstractMojo {
     @Parameter(name = "yarnVersion", defaultValue = "v1.6.0", required = true)
     private String yarnVersion;
 
+    @Parameter(name = "yarnNetworkConcurrency", defaultValue = "-1")
+    private int yarnNetworkConcurrency;
+
     @Parameter(property = "ignoreMavenProxies", defaultValue = "true", required = true)
     private boolean ignoreMavenProxies;
 
@@ -113,9 +116,9 @@ public class PackageForProductionMojo extends AbstractMojo {
                 transpileWorkingDirectory, es5OutputDirectoryName,
                 es6OutputDirectoryName, frontendDataProvider);
         new TranspilationStep(frontendToolsManager, getProxyConfig(),
-                nodeVersion, yarnVersion).transpileFiles(
-                        transpileEs6SourceDirectory, transpileOutputDirectory,
-                        skipEs5);
+                nodeVersion, yarnVersion, yarnNetworkConcurrency)
+                        .transpileFiles(transpileEs6SourceDirectory,
+                                transpileOutputDirectory, skipEs5);
     }
 
     private Map<String, Set<String>> getFragmentsData(

--- a/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/production/TranspilationStep.java
+++ b/flow-maven-plugin/src/main/java/com/vaadin/flow/plugin/production/TranspilationStep.java
@@ -22,13 +22,13 @@ import java.util.Map;
 import java.util.Objects;
 
 import com.github.eirslett.maven.plugins.frontend.lib.ProxyConfig;
-
 import com.vaadin.flow.plugin.common.FrontendToolsManager;
 
 /**
  * Transpiles artifacts in the specified directory.
  * <p>
- * Note: this class is intended to be independent from Maven dependencies so that it can be reused in Gradle plugin in future.
+ * Note: this class is intended to be independent from Maven dependencies so
+ * that it can be reused in Gradle plugin in future.
  *
  * @author Vaadin Ltd.
  */
@@ -38,35 +38,55 @@ public class TranspilationStep {
     /**
      * Prepares the step.
      *
-     * @param frontendToolsManager the manager to be used to transpile files, not {@code null}
-     * @param proxyConfig proxy config to use when downloading frontend tools, not {@code null}
-     * @param nodeVersion node version to install, not {@code null}
-     * @param yarnVersion yarn version to install, not {@code null}
+     * @param frontendToolsManager
+     *            the manager to be used to transpile files, not {@code null}
+     * @param proxyConfig
+     *            proxy config to use when downloading frontend tools, not
+     *            {@code null}
+     * @param nodeVersion
+     *            node version to install, not {@code null}
+     * @param yarnVersion
+     *            yarn version to install, not {@code null}
+     * @param networkConcurrency
+     *            maximum number of concurrent network requests
      */
-    public TranspilationStep(FrontendToolsManager frontendToolsManager, ProxyConfig proxyConfig, String nodeVersion, String yarnVersion) {
-        this.frontendToolsManager = Objects.requireNonNull(frontendToolsManager);
-        frontendToolsManager.installFrontendTools(proxyConfig, nodeVersion, yarnVersion);
+    public TranspilationStep(FrontendToolsManager frontendToolsManager,
+            ProxyConfig proxyConfig, String nodeVersion, String yarnVersion,
+            int networkConcurrency) {
+        this.frontendToolsManager = Objects
+                .requireNonNull(frontendToolsManager);
+        frontendToolsManager.installFrontendTools(proxyConfig, nodeVersion,
+                yarnVersion, networkConcurrency);
     }
 
     /**
      * Transpiles the files from source directory into the output directory.
      *
-     * @param es6SourceDirectory   the directory with original ES6 files
-     * @param outputDirectory      the directory that will have processed files in
-     * @param skipEs5              whether to skip the transpilation step or not
-     * @throws IllegalStateException if no transpilation results found
-     * @throws UncheckedIOException  if {@link IOException} occurs during file operations
+     * @param es6SourceDirectory
+     *            the directory with original ES6 files
+     * @param outputDirectory
+     *            the directory that will have processed files in
+     * @param skipEs5
+     *            whether to skip the transpilation step or not
+     * @throws IllegalStateException
+     *             if no transpilation results found
+     * @throws UncheckedIOException
+     *             if {@link IOException} occurs during file operations
      */
-    public void transpileFiles(File es6SourceDirectory, File outputDirectory, boolean skipEs5) {
-        Map<String, File> transpilationResult = frontendToolsManager.transpileFiles(es6SourceDirectory, outputDirectory, skipEs5);
+    public void transpileFiles(File es6SourceDirectory, File outputDirectory,
+            boolean skipEs5) {
+        Map<String, File> transpilationResult = frontendToolsManager
+                .transpileFiles(es6SourceDirectory, outputDirectory, skipEs5);
         if (transpilationResult.isEmpty()) {
-            throw new IllegalStateException("Received no transpilation results from frontend tools");
+            throw new IllegalStateException(
+                    "Received no transpilation results from frontend tools");
         }
-        transpilationResult.values().stream()
-                .filter(configurationOutput -> !configurationOutput.isDirectory())
-                .findAny()
-                .ifPresent(nonExistingDirectory -> {
-                    throw new IllegalStateException(String.format("Transpilation output at '%s' is not a directory or does not exist", nonExistingDirectory));
+        transpilationResult.values().stream().filter(
+                configurationOutput -> !configurationOutput.isDirectory())
+                .findAny().ifPresent(nonExistingDirectory -> {
+                    throw new IllegalStateException(String.format(
+                            "Transpilation output at '%s' is not a directory or does not exist",
+                            nonExistingDirectory));
                 });
     }
 }

--- a/flow-maven-plugin/src/test/java/com/vaadin/flow/plugin/production/TranspilationStepTest.java
+++ b/flow-maven-plugin/src/test/java/com/vaadin/flow/plugin/production/TranspilationStepTest.java
@@ -1,22 +1,5 @@
 package com.vaadin.flow.plugin.production;
 
-import java.io.File;
-import java.io.IOException;
-import java.io.UncheckedIOException;
-import java.util.Collections;
-import java.util.List;
-
-import com.github.eirslett.maven.plugins.frontend.lib.ProxyConfig;
-import com.google.common.collect.ImmutableMap;
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.rules.ExpectedException;
-import org.junit.rules.TemporaryFolder;
-
-import com.vaadin.flow.plugin.TestUtils;
-import com.vaadin.flow.plugin.common.FrontendDataProvider;
-import com.vaadin.flow.plugin.common.FrontendToolsManager;
-
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
@@ -26,6 +9,23 @@ import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Collections;
+import java.util.List;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+
+import com.github.eirslett.maven.plugins.frontend.lib.ProxyConfig;
+import com.google.common.collect.ImmutableMap;
+import com.vaadin.flow.plugin.TestUtils;
+import com.vaadin.flow.plugin.common.FrontendDataProvider;
+import com.vaadin.flow.plugin.common.FrontendToolsManager;
 
 /**
  * @author Vaadin Ltd.
@@ -40,15 +40,23 @@ public class TranspilationStepTest {
     private final boolean skipEs5 = true;
     private final String nodeVersion = "1";
     private final String yarnVersion = "1";
-    private final ProxyConfig proxyConfig = new ProxyConfig(Collections.emptyList());
+    private final int networkConcurrency = 1;
+    private final ProxyConfig proxyConfig = new ProxyConfig(
+            Collections.emptyList());
 
     private FrontendToolsManager getFrontendToolsManager(File outputDirectory) {
-        FrontendDataProvider frontendDataProviderMock = mock(FrontendDataProvider.class);
-        when(frontendDataProviderMock.createShellFile(outputDirectory)).thenReturn("shell");
-        when(frontendDataProviderMock.createFragmentFiles(outputDirectory)).thenReturn(Collections.singleton("fragment"));
+        FrontendDataProvider frontendDataProviderMock = mock(
+                FrontendDataProvider.class);
+        when(frontendDataProviderMock.createShellFile(outputDirectory))
+                .thenReturn("shell");
+        when(frontendDataProviderMock.createFragmentFiles(outputDirectory))
+                .thenReturn(Collections.singleton("fragment"));
 
-        FrontendToolsManager frontendToolsManager = spy(new FrontendToolsManager(outputDirectory, "frontend-es5", "frontend-es6", frontendDataProviderMock));
-        doNothing().when(frontendToolsManager).installFrontendTools(proxyConfig, nodeVersion, yarnVersion);
+        FrontendToolsManager frontendToolsManager = spy(
+                new FrontendToolsManager(outputDirectory, "frontend-es5",
+                        "frontend-es6", frontendDataProviderMock));
+        doNothing().when(frontendToolsManager).installFrontendTools(proxyConfig,
+                nodeVersion, yarnVersion, networkConcurrency);
         return frontendToolsManager;
     }
 
@@ -58,8 +66,10 @@ public class TranspilationStepTest {
         exception.expect(UncheckedIOException.class);
         exception.expectMessage(fileNotDirectory.toString());
 
-        new TranspilationStep(getFrontendToolsManager(temporaryFolder.getRoot()), proxyConfig, nodeVersion, yarnVersion)
-                .transpileFiles(temporaryFolder.getRoot(), fileNotDirectory, skipEs5);
+        new TranspilationStep(
+                getFrontendToolsManager(temporaryFolder.getRoot()), proxyConfig,
+                nodeVersion, yarnVersion, networkConcurrency).transpileFiles(
+                        temporaryFolder.getRoot(), fileNotDirectory, skipEs5);
     }
 
     @Test
@@ -68,8 +78,10 @@ public class TranspilationStepTest {
         exception.expect(IllegalArgumentException.class);
         exception.expectMessage(nonExistingFile.toString());
 
-        new TranspilationStep(getFrontendToolsManager(temporaryFolder.getRoot()), proxyConfig, nodeVersion, yarnVersion)
-                .transpileFiles(nonExistingFile, temporaryFolder.getRoot(), skipEs5);
+        new TranspilationStep(
+                getFrontendToolsManager(temporaryFolder.getRoot()), proxyConfig,
+                nodeVersion, yarnVersion, networkConcurrency).transpileFiles(
+                        nonExistingFile, temporaryFolder.getRoot(), skipEs5);
     }
 
     @Test
@@ -78,25 +90,33 @@ public class TranspilationStepTest {
         exception.expect(IllegalArgumentException.class);
         exception.expectMessage(fileNotDirectory.toString());
 
-        new TranspilationStep(getFrontendToolsManager(temporaryFolder.getRoot()), proxyConfig, nodeVersion, yarnVersion)
-                .transpileFiles(fileNotDirectory, temporaryFolder.getRoot(), skipEs5);
+        new TranspilationStep(
+                getFrontendToolsManager(temporaryFolder.getRoot()), proxyConfig,
+                nodeVersion, yarnVersion, networkConcurrency).transpileFiles(
+                        fileNotDirectory, temporaryFolder.getRoot(), skipEs5);
     }
 
     @Test
     public void transpileFiles_emptyResults() {
         File outputDirectory = new File("output");
         File es6Directory = new File(outputDirectory, "frontend");
-        FrontendToolsManager toolsManagerMock = mock(FrontendToolsManager.class);
-        when(toolsManagerMock.transpileFiles(es6Directory, outputDirectory, skipEs5)).thenReturn(Collections.emptyMap());
+        FrontendToolsManager toolsManagerMock = mock(
+                FrontendToolsManager.class);
+        when(toolsManagerMock.transpileFiles(es6Directory, outputDirectory,
+                skipEs5)).thenReturn(Collections.emptyMap());
 
         try {
-            new TranspilationStep(toolsManagerMock, proxyConfig, nodeVersion, yarnVersion).transpileFiles(es6Directory, outputDirectory, skipEs5);
+            new TranspilationStep(toolsManagerMock, proxyConfig, nodeVersion,
+                    yarnVersion, networkConcurrency).transpileFiles(
+                            es6Directory, outputDirectory, skipEs5);
             fail("Frontend manager had returned empty transpilation results, but the step does not fail");
         } catch (IllegalStateException expected) {
-            //expected
+            // expected
         }
-        verify(toolsManagerMock, times(1)).installFrontendTools(proxyConfig, nodeVersion, yarnVersion);
-        verify(toolsManagerMock, times(1)).transpileFiles(es6Directory, outputDirectory, skipEs5);
+        verify(toolsManagerMock, times(1)).installFrontendTools(proxyConfig,
+                nodeVersion, yarnVersion, networkConcurrency);
+        verify(toolsManagerMock, times(1)).transpileFiles(es6Directory,
+                outputDirectory, skipEs5);
     }
 
     @Test
@@ -104,45 +124,74 @@ public class TranspilationStepTest {
         File outputDirectory = temporaryFolder.newFolder("output");
         File es6Directory = temporaryFolder.newFolder("output", "frontend");
         File nonExistingFile = new File("doesNotExist");
-        FrontendToolsManager toolsManagerMock = mock(FrontendToolsManager.class);
-        when(toolsManagerMock.transpileFiles(es6Directory, outputDirectory, skipEs5)).thenReturn(Collections.singletonMap(es6Directory.getName(), nonExistingFile));
+        FrontendToolsManager toolsManagerMock = mock(
+                FrontendToolsManager.class);
+        when(toolsManagerMock.transpileFiles(es6Directory, outputDirectory,
+                skipEs5)).thenReturn(
+                        Collections.singletonMap(es6Directory.getName(),
+                                nonExistingFile));
 
         try {
-            new TranspilationStep(toolsManagerMock, proxyConfig, nodeVersion, yarnVersion).transpileFiles(es6Directory, outputDirectory, skipEs5);
-            fail(String.format("Directory '%s' does not contain transpilation results, but the step does not fail", nonExistingFile));
+            new TranspilationStep(toolsManagerMock, proxyConfig, nodeVersion,
+                    yarnVersion, networkConcurrency).transpileFiles(
+                            es6Directory, outputDirectory, skipEs5);
+            fail(String.format(
+                    "Directory '%s' does not contain transpilation results, but the step does not fail",
+                    nonExistingFile));
         } catch (IllegalStateException expected) {
-            //expected
+            // expected
         }
-        verify(toolsManagerMock, times(1)).installFrontendTools(proxyConfig, nodeVersion, yarnVersion);
-        verify(toolsManagerMock, times(1)).transpileFiles(es6Directory, outputDirectory, skipEs5);
+        verify(toolsManagerMock, times(1)).installFrontendTools(proxyConfig,
+                nodeVersion, yarnVersion, networkConcurrency);
+        verify(toolsManagerMock, times(1)).transpileFiles(es6Directory,
+                outputDirectory, skipEs5);
     }
 
     @Test
     public void transpileFiles() throws IOException {
         File outputDirectory = temporaryFolder.newFolder("target");
-        File es6SourceDirectory = temporaryFolder.newFolder("target", "frontend");
+        File es6SourceDirectory = temporaryFolder.newFolder("target",
+                "frontend");
         temporaryFolder.newFile("target/frontend/index-es6-original.html");
 
-        List<String> sourceFiles = TestUtils.listFilesRecursively(es6SourceDirectory);
-        File es5TranspiledDirectory = temporaryFolder.newFolder("target", "build", "frontend-es5");
-        File es5TranspiledFile = temporaryFolder.newFile("target/build/frontend-es5/index-es5.html");
+        List<String> sourceFiles = TestUtils
+                .listFilesRecursively(es6SourceDirectory);
+        File es5TranspiledDirectory = temporaryFolder.newFolder("target",
+                "build", "frontend-es5");
+        File es5TranspiledFile = temporaryFolder
+                .newFile("target/build/frontend-es5/index-es5.html");
 
-        File es6TranspiledDirectory = temporaryFolder.newFolder("target", "build", "frontend-es6");
-        File es6transpiledFile = temporaryFolder.newFile("target/build/frontend-es6/index-es6.html");
+        File es6TranspiledDirectory = temporaryFolder.newFolder("target",
+                "build", "frontend-es6");
+        File es6transpiledFile = temporaryFolder
+                .newFile("target/build/frontend-es6/index-es6.html");
 
-        FrontendToolsManager toolsManagerMock = mock(FrontendToolsManager.class);
-        when(toolsManagerMock.transpileFiles(es6SourceDirectory, outputDirectory, skipEs5))
-                .thenReturn(ImmutableMap.of("frontend-es5", es5TranspiledDirectory, "frontend-es6", es6TranspiledDirectory));
+        FrontendToolsManager toolsManagerMock = mock(
+                FrontendToolsManager.class);
+        when(toolsManagerMock.transpileFiles(es6SourceDirectory,
+                outputDirectory, skipEs5)).thenReturn(
+                        ImmutableMap.of("frontend-es5", es5TranspiledDirectory,
+                                "frontend-es6", es6TranspiledDirectory));
 
-        new TranspilationStep(toolsManagerMock, proxyConfig, nodeVersion, yarnVersion).transpileFiles(es6SourceDirectory, outputDirectory, skipEs5);
+        new TranspilationStep(toolsManagerMock, proxyConfig, nodeVersion,
+                yarnVersion, networkConcurrency).transpileFiles(
+                        es6SourceDirectory, outputDirectory, skipEs5);
 
-        assertEquals("Es6 source files should be left untouched", sourceFiles, TestUtils.listFilesRecursively(es6SourceDirectory));
+        assertEquals("Es6 source files should be left untouched", sourceFiles,
+                TestUtils.listFilesRecursively(es6SourceDirectory));
 
-        List<String> pathsAfterTranspilation = TestUtils.listFilesRecursively(outputDirectory);
-        assertTrue("ES5 transpilation result should be present", pathsAfterTranspilation.stream().anyMatch(path1 -> path1.endsWith(es5TranspiledFile.getName())));
-        assertTrue("ES6 transpilation result should be present", pathsAfterTranspilation.stream().anyMatch(path1 -> path1.endsWith(es6transpiledFile.getName())));
+        List<String> pathsAfterTranspilation = TestUtils
+                .listFilesRecursively(outputDirectory);
+        assertTrue("ES5 transpilation result should be present",
+                pathsAfterTranspilation.stream().anyMatch(
+                        path1 -> path1.endsWith(es5TranspiledFile.getName())));
+        assertTrue("ES6 transpilation result should be present",
+                pathsAfterTranspilation.stream().anyMatch(
+                        path1 -> path1.endsWith(es6transpiledFile.getName())));
 
-        verify(toolsManagerMock, times(1)).installFrontendTools(proxyConfig, nodeVersion, yarnVersion);
-        verify(toolsManagerMock, times(1)).transpileFiles(es6SourceDirectory, outputDirectory, skipEs5);
+        verify(toolsManagerMock, times(1)).installFrontendTools(proxyConfig,
+                nodeVersion, yarnVersion, networkConcurrency);
+        verify(toolsManagerMock, times(1)).transpileFiles(es6SourceDirectory,
+                outputDirectory, skipEs5);
     }
 }

